File: scripts/dev-server.py
#!/usr/bin/env python3
"""
Local dev server that mimics Cloudflare's clean-URL asset resolution
(html_handling: "auto-trailing-slash" in wrangler.jsonc), since plain
`python3 -m http.server` only serves exact file paths and 404s on the
extension-less URLs this site actually links (e.g. /games/math/foo).

  /                       -> index.html
  /games/math/foo         -> serves games/math/foo.html directly (200)
  /games/math/foo.html    -> 307 redirect to /games/math/foo, matching
                             production's own canonicalization
  /index.html             -> 307 redirect to /

Responses are sent with no-cache headers — unlike production's
_headers file, local dev should always show your latest edit, not
whatever cache duration Cloudflare would apply.

Usage: python3 scripts/dev-server.py [port]   (default port 8000)
"""
import http.server
import os
import sys
from urllib.parse import urlsplit

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PORT = int(sys.argv[1]) if len(sys.argv) > 1 else 8000


def load_assetsignore():
    """Same file Cloudflare reads to decide what NOT to deploy/serve —
    read it here too so local dev 404s on the same paths production does
    (repo/tooling files like .git, README.md, wrangler.jsonc)."""
    patterns = []
    path = os.path.join(ROOT, '.assetsignore')
    if os.path.isfile(path):
        with open(path) as f:
            for line in f:
                line = line.strip().strip('/')
                if line and not line.startswith('#'):
                    patterns.append(line)
    return patterns


IGNORED = load_assetsignore()


def is_ignored(path):
    parts = path.lstrip('/').split('/')
    return any('/'.join(parts[:i]) in IGNORED for i in range(1, len(parts) + 1))


class CleanUrlHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def send_head(self):
        parsed = urlsplit(self.path)
        path = parsed.path
        query_suffix = ('?' + parsed.query) if parsed.query else ''

        if is_ignored(path):
            self.send_error(404, "File not found")
            return None

        # Literal .html request -> redirect to the clean URL, same as
        # production. /index.html is a special case: it collapses to /
        # rather than /index.
        if path.endswith('.html'):
            clean = path[:-len('.html')]
            if clean.endswith('/index'):
                clean = clean[:-len('index')]
            self.send_response(307)
            self.send_header('Location', clean + query_suffix)
            self.end_headers()
            return None

        # Clean URL -> serve the matching .html file directly (no
        # redirect), if one exists and nothing already matches literally.
        if path != '/' and not path.endswith('/'):
            literal = self.translate_path(path)
            if not os.path.isfile(literal):
                candidate = self.translate_path(path + '.html')
                if os.path.isfile(candidate):
                    self.path = path + '.html' + query_suffix

        return super().send_head()

    def end_headers(self):
        self.send_header('Cache-Control', 'no-cache, no-store, must-revalidate')
        super().end_headers()


# A browser loads a page's CSS/JS/images as several requests at once
# and cancels others (favicons, aborted navigations); the plain
# single-threaded server in the stdlib serves those serially and
# prints a full traceback for every cancelled one. ThreadingHTTPServer
# handles them concurrently, and handle_error() below silences the
# ones that are just "the browser closed the connection," not a bug.
class DevServer(http.server.ThreadingHTTPServer):
    def handle_error(self, request, client_address):
        exc_type = sys.exc_info()[0]
        if exc_type and issubclass(
            exc_type, (BrokenPipeError, ConnectionResetError, ConnectionAbortedError)
        ):
            return
        super().handle_error(request, client_address)


if __name__ == '__main__':
    with DevServer(('', PORT), CleanUrlHandler) as httpd:
        print(f"Serving {ROOT}")
        print(f"http://localhost:{PORT}/  (clean URLs, no caching)")
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            pass

File: scripts/dev-server.js
#!/usr/bin/env node
'use strict';
/*
 * Local dev server that mimics Cloudflare's clean-URL asset resolution
 * (html_handling: "auto-trailing-slash" in wrangler.jsonc), since a
 * plain static file server only serves exact file paths and 404s on
 * the extension-less URLs this site actually links (e.g. /games/math/foo).
 *
 *   /                       -> index.html
 *   /games/math/foo         -> serves games/math/foo.html directly (200)
 *   /games/math/foo.html    -> 307 redirect to /games/math/foo, matching
 *                              production's own canonicalization
 *   /index.html             -> 307 redirect to /
 *
 * Responses are sent with no-cache headers — unlike production's
 * _headers file, local dev should always show your latest edit, not
 * whatever cache duration Cloudflare would apply.
 *
 * Usage: node scripts/dev-server.js [port]   (default port 8000)
 */
const http = require('node:http');
const fs = require('node:fs');
const path = require('node:path');

const ROOT = path.dirname(__dirname);
const PORT = Number(process.argv[2]) || 8000;

const MIME_TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.png': 'image/png',
  '.svg': 'image/svg+xml',
  '.ico': 'image/x-icon',
  '.txt': 'text/plain; charset=utf-8',
};

function loadAssetsignore() {
  const file = path.join(ROOT, '.assetsignore');
  if (!fs.existsSync(file)) return [];
  return fs
    .readFileSync(file, 'utf8')
    .split('\n')
    .map((line) => line.trim().replace(/^\/+|\/+$/g, ''))
    .filter((line) => line && !line.startsWith('#'));
}

const IGNORED = loadAssetsignore();

function isIgnored(urlPath) {
  const parts = urlPath.replace(/^\/+/, '').split('/');
  return parts.some((_part, i) => IGNORED.includes(parts.slice(0, i + 1).join('/')));
}

// Resolve a URL path to a file under ROOT, refusing to escape it via `..`.
function resolveSafe(urlPath) {
  const resolved = path.normalize(path.join(ROOT, decodeURIComponent(urlPath)));
  if (resolved !== ROOT && !resolved.startsWith(ROOT + path.sep)) return null;
  return resolved;
}

function isFile(filePath) {
  try {
    return fs.statSync(filePath).isFile();
  } catch {
    return false;
  }
}

function serveFile(res, filePath) {
  const type = MIME_TYPES[path.extname(filePath).toLowerCase()] || 'application/octet-stream';
  res.writeHead(200, {
    'Content-Type': type,
    'Cache-Control': 'no-cache, no-store, must-revalidate',
  });
  fs.createReadStream(filePath).on('error', () => res.destroy()).pipe(res);
}

function notFound(res) {
  res.writeHead(404, {
    'Content-Type': 'text/plain; charset=utf-8',
    'Cache-Control': 'no-cache, no-store, must-revalidate',
  });
  res.end('404 Not Found');
}

const server = http.createServer((req, res) => {
  // A browser cancels some requests (aborted navigations, favicons);
  // that's not a real server error, just the client hanging up early.
  req.on('error', () => {});
  res.on('error', (err) => {
    if (err.code !== 'EPIPE' && err.code !== 'ECONNRESET') console.error(err);
  });

  const url = new URL(req.url, `http://${req.headers.host}`);
  const urlPath = decodeURIComponent(url.pathname);

  if (isIgnored(urlPath)) return notFound(res);

  // Literal .html request -> redirect to the clean URL, same as
  // production. /index.html is a special case: it collapses to /
  // rather than /index.
  if (urlPath.endsWith('.html')) {
    let clean = urlPath.slice(0, -'.html'.length);
    if (clean.endsWith('/index')) clean = clean.slice(0, -'index'.length);
    res.writeHead(307, { Location: clean + url.search });
    return res.end();
  }

  if (urlPath === '/' || urlPath.endsWith('/')) {
    const indexPath = resolveSafe(urlPath + 'index.html');
    if (indexPath && isFile(indexPath)) return serveFile(res, indexPath);
    return notFound(res);
  }

  // Exact match (css/js/images/etc) -> serve as-is.
  const literal = resolveSafe(urlPath);
  if (literal && isFile(literal)) return serveFile(res, literal);

  // Clean URL -> serve the matching .html file directly (no redirect).
  const htmlCandidate = resolveSafe(urlPath + '.html');
  if (htmlCandidate && isFile(htmlCandidate)) return serveFile(res, htmlCandidate);

  return notFound(res);
});

server.on('error', (err) => {
  if (err.code === 'EADDRINUSE') {
    console.error(`Port ${PORT} is already in use — pass a different port: node scripts/dev-server.js <port>`);
    process.exit(1);
  }
  throw err;
});

server.listen(PORT, () => {
  console.log(`Serving ${ROOT}`);
  console.log(`http://localhost:${PORT}/  (clean URLs, no caching)`);
});
